debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/events/BufferingChangeStreamCursor.java

```diff
@@ -0,0 +1,303 @@
+        }
+
+        public boolean isEmpty() {
+            return queue.isEmpty();
+        }
+
+        /**
+         * Depending on queue implementation this method may not be reliable
+         * By default see {@link ConcurrentLinkedQueue#size()}
+         *
+         * @return approximate number of elements in queue
+         */
+        public int size() {
+            return queue.size();
+        }
+
+        @Override
+        public void run() {
+            try (MongoChangeStreamCursor<ChangeStreamDocument<TResult>> cursor = stream.cursor()) {
+                cursorRef.compareAndSet(null, cursor);
+                running.set(true);
+                noMessageIterations = 0;
+                fetchEvents(cursor);
+            }
+            catch (InterruptedException e) {
+                throw new DebeziumException("Fetcher thread interrupted", e);
+            }
+            finally {
+                cursorRef.set(null);
+                close();
+            }
+        }
+
+        private void fetchEvents(MongoChangeStreamCursor<ChangeStreamDocument<TResult>> cursor) throws InterruptedException {
+            while (isRunning()) {
+                var beforeEventPollTime = clock.currentTimeAsInstant();
+                var event = new ResumableChangeStreamEvent<>(cursor.next(), cursor);
+                event.document.ifPresent(doc -> LOGGER.trace("Polled Change Stream event: {}", event));
+                metrics.onSourceEventPolled(event, clock, beforeEventPollTime);
+
+                if (event.hasDocument()) {
+                    enqueue(event);
+                }
+                else {
+                    throttleIfNeeded(event);
+                }
+            }
+        }
+
+        private void throttleIfNeeded(ResumableChangeStreamEvent<TResult> event) throws InterruptedException {
+            noMessageIterations++;
+            if (noMessageIterations >= THROTTLE_NO_MESSAGE_BEFORE_PAUSE) {
+                enqueue(event);
+                noMessageIterations = 0;
+                pause.sleepWhen(true);
+            }
+        }
+
+        private void enqueue(ResumableChangeStreamEvent<TResult> event) throws InterruptedException {
+            this.capacity.acquire();
+            queue.offer(event);
+        }
+    }
+
+    public static <TResult> BufferingChangeStreamCursor<TResult> fromIterable(
+                                                                              ChangeStreamIterable<TResult> stream,
+                                                                              MongoDbTaskContext taskContext,
+                                                                              MongoDbStreamingChangeEventSourceMetrics metrics,
+                                                                              Clock clock) {
+        var pause = DelayStrategy.constant(taskContext.getConnectorConfig().getPollInterval());
+        var fetcher = new EventFetcher<>(stream, 1000, pause, metrics, clock);
```

Comment:
Do we want to reuse some config property or introduce a new one ?

Reply:
What about we do introduce a new config (e.g. `cursor.max.batch.size`) defaulting either to zero or the value of `snapshot.fetch.size`  and then have the queue sized as  `min(3*cursor.max.batch.size, max.batch.size)`.  The obvious problem is that the default cursor.max.batch.size is 0 which means server default (usually as "I'l try fill 16MB")

So maybe we could just set it to `max.batch.size`. I think we should be careful here as this increases the required memory. So in a sense standalone config property defaulting to `max.batch.size` is really the optimal, however I understand the need not to increase the already overwhelming number of configuration properties we have :(

So maybe reusing the `max.batch.size` is really the way to go here

Reply:
Well max.batech.size is always <= max.queue.size

Reply:
I see, we ensure this in validation. So a value of `max.batch.size` sounds as a good start I think.

Reply:
That also makes quite predictable what is the maximum number of records/events the connector can keep in memory during runtime. `max.queue.size` of records + `max.batch.size` resumable events. In reality this maximum is never going to be achieved unless KC consumer takes a vacation.